x-pack/plugins/maps/public/connected_components/mb_map/tile_status_tracker.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Map as MapboxMap, MapDataEvent } from 'mapbox-gl';
+import _ from 'lodash';
+import { ILayer } from '../../classes/layers/layer';
+import { SPATIAL_FILTERS_LAYER_ID } from '../../../common/constants';
+
+interface MbTile {
+  // references internal object from mapbox
+  aborted?: true;
+}
+
+interface Tile {
+  mbKey: string;
+  mbSourceId: string;
+  mbTile: MbTile;
+}
+
+export class TileStatusTracker {
+  private _tileCache: Tile[];
+  private readonly _mbMap: MapboxMap;
+  private readonly _setAreTilesLoaded: (layerId: string, areTilesLoaded: boolean) => void;
+  private readonly _getCurrentLayerList: () => ILayer[];
+  private readonly _onSourceDataLoading = (e: MapDataEvent & { sourceId: string }) => {
+    if (
+      e.sourceId &&
+      e.sourceId !== SPATIAL_FILTERS_LAYER_ID &&
+      e.dataType === 'source' &&
+      e.tile
+    ) {
+      const tracked = this._tileCache.find((tile) => {
+        return (
+          tile.mbKey === ((e.coord.key as unknown) as string) && tile.mbSourceId === e.sourceId
```

Comment:
`_onSourceDataLoading` and `_onSourceData` get layer id from a different location then `_onError`. To be consistent, how about always pulling mbKey from `e.tile.tileID.key` like in `_onError`?